### ISSUE TITLE
Backup: Do not exclude /data/media

### ISSUE THREAD

**Issue by leo-bogert** [TeamWin/Team-Win-Recovery-Project#276]
When doing a backup on my Galaxy Nexus, TWRP 2.7.0.0 will exclude anything in /data/media because it puts its backup in /data/media/TWRP (or /data/media/0/TWRP).

But /data/media typically contains a lot of user data, especially the camera pictures.
Further, lots of Android apps have the bad programming style of storing their data there, for example WhatsApp.

This renders the TWRP backups rather incomplete. Can you please change it to only exclude its own backup folder and include everything in /data/media?

**Comment by zeorin** [comment]
+1

I know I've seen that the prevailing opinion has been that there are other ways to back this up (e.g. just copy stuff over onto a PC), but many apps do indeed store their stuff in that folder.

Also, it's not very obvious to a user that when they're selecting the DATA partition for backup, that it's not actually the DATA partition that's backed up, just parts of it.

I myself never employed any of the 'alternative backup methods' for /data/media because I was (reasonably, IMO) under the impression that I was often backing that up when I selected the DATA partition.

I've now lost many sentimentally valuable photographs. I know people say you should back that stuff up, but _that's what I thought I was doing_.

So, either include /data/media (except other backups), or make it _very clear_ what is actually backed up when someone selects the DATA partition for backup.

**Comment by tdussmann** [comment]
Def. +1

I was just wondering whether this was the case for a backup of the data partition or now. 
Sure, when creating a backup of /data onto /data/media it would create a loop, 
but for all other targets, like external SD-Card or USB OTG, it would be possible to copy the 'external storage' as well...

It could be shown only when selecting a supported backup location and to make it clear, that media/ is not included, it could be greyed out when backing up to it...

So, yeah, we can still take a full backup by backing up storage manually after a TWRP backup, but for a full blown backup/restore solution, it should be able to support this. After all, a lot of people use TWRP for it's extensive functionality.

**Comment by prataczak** [comment]
I was actually just getting ready to flash Odin, with the intention of flashing back to Cyanogenmod 12.1, but on a hunch, I decided to research this issue (because I remembered my last restore said something about excluding data/media).

I guess it's a good thing I did, seeing as flashing the stock ROM will result in being stuck in boot loop unless I do a factory reset on my Samsung Galaxy S4.

I'm new to using TWRP, and I have to say I already like it better than CWM, for the simple fact that _every single device_ I have used CWM on has resulted in failed restores and hours of frustration manually restoring apps.

But this right here... this defies logic. This is not intuitive at all. With all due respect... who honestly decided it would be a good idea to skip backup of a folder that contains tons of data (I just checked mine - it had 1.6 gb of app data)?

And I never would have thought to 'manually' back up that folder. I shouldn't have to. I run all my backups to an external SD card just because I am paranoid about losing anything.

This should be fixed. Once again... this defies logic.

**Comment by smemsh** [comment]
Just lost a bunch of important files because of this on my `titan` board.  Thought I _was_ backing it up because:
1. I selected to backup `/data`,
2. there _was_ a "microsd" checkbox (external sdcard), but
3. there _was not_ any checkbox for internal sdcard
4. I presumed the list was exhaustive (and would backup everything if all were checked)

so my assumption was that internal sdcard contents were included in `/data` (since it actually is within `/data` filesystem in `/data/media/0` this seems like a reasonable assumption)

definitely at least the user should be warned, I would have backed it up manually had I known.  Only after wiping and restoring, but having all kinds of errors coming back up (even google apps are storing things here) did I inspect the tar file and find that entire section was missing!

**Comment by biji** [comment]
Yeah ... Dev please implement this, lost my media files, I thought backup /data is enough

**Comment by kdrag0n** [comment]
+1

Lost all my game progress, documents, and photos. At least warn that the internal storage will not be backed up. Maybe add another option to back up "Internal Storage" (the term used for /data/media in the rest of TWRP)

Please fix, TWRP.

**Comment by niker** [comment]
On android 5.x the problem is even worse, since the access to this internal storage is limited via permissions/SELinux for all apps. I copied all contents via ADB as a backup, but folders created by an app must have the correct permission/context set AFTER the restore or the app can only access said folder in read-only mode. This means even though I did backup to PC, the backup was worthless and after restore half of my apps kept crashing or behave strangely due to permission problems. Backing this up conveniently in recovery should definitely be an option.

**Comment by lchiocca** [comment]
This would be a really nice addition. But I have to admit that this would only be a use-case when backing up to an external device (via usb-otg or an external sd-card). I've been looking into the code and it doesn't seem easy to "pull up" this parameter to the UI. As a quick workaround, would adding a global variable that would override this behaviour be an option (e.g. TW_FORCE_BACKUP_OF_MEDIA_DATA)?

**Comment by JustMyGithub** [comment]
Another +1 from me.
surely you can backup it to a connected computer, but:
a) MTP is slow for a large number of files (much faster for an archive)
b) application data is stored on internal storage (you backup android.secure, don't you? - treat app data on internal storage differently)

I just recently lost app data due to a bug in the app. If the backup would have included media folder, I could restore the app data now....

Of course you better exclude the default TWRP and CWM backup paths ;D

**Comment by wdormann** [comment]
As a workaround, I find it quite reasonable to rsync the sdcard periodically.  Being rsync, the incremental backups are super fast.   And the combination of a restore from a nandroid backup plus a reverse rsync will put you right back where you were.    https://play.google.com/store/apps/details?id=eu.kowalczuk.rsync4android&hl=en

**Comment by eulerreich** [comment]
+1 for me as well. Lost a lot of data because I thought TWRP got my back on /data/media

**Comment by leo-bogert** [comment]
Can someone please fix this ASAP?
I need this really urgently. Please help.

**Comment by leo-bogert** [comment]
Or even more easy:
Can someone please merge the fix for this?
Github automatically listed some commits above from another repository which seem to fix this.
Please, I need this ASAP, it would be of great help to me :)

**Comment by CaptainThrowback** [comment]
No one has submitted any patches to gerrit for this.

**Comment by leo-bogert** [comment]
I'm willing to pay a $ 15 bounty in Bitcoin to whoever fixes this.

Please prove it by linking the merged pull request with your fix.

**EDIT**: Bounty currently not available anymore, sorry.

**Comment by cmorlok** [comment]
Some information regarding /data/media: https://twrp.me/faq/datamedia.html

**Comment by blackwind** [comment]
That this is still an issue over two years later is just plain irresponsible on the part of the developers. I was fortunate that I had just begun my Android journey when first afflicted. Others here, apparently, weren't so lucky.

As noted by others above, if you're not going to include `/data/media`, at least inform the user in no uncertain terms. But you _should_ include `/data/media`. That's what 100% of your users are expecting, and no one should have to learn such an important lesson the hard way.

**Comment by kitambi** [comment]
100% agree here.  I probably should have read this (two year old) issue before trying this, but let's just sum it up here:
All of my photos from the last several months are gone because TWRP failed to back them up.
I have Titanium Backup.
My phone has an external SD card.
I don't give a crap about anything in /data except...oh yeah, /data/media.

So, thanks for that.  Every once in a while, I consider just dumping Android completely and getting a flip phone.  This is one of those times.

Note: FIX THIS. NOW. PEOPLE STORE IMPORTANT THINGS IN /data/media!  I'm afraid that I can't recommend TWRP to anyone in good faith until this is fixed.

**Comment by blackwind** [comment]
Tagging the most active contributors since this issue seems to be buried deep: @Dees-Troy @that1 @bigbiff @mdmower @Tasssadar @enh @z31s1g @andi34

**Comment by JustMyGithub** [comment]
what about using a separate file for /data/media-backups.
that way people can unselect or choose to backup the internal sdcard easily.

**Comment by kitambi** [comment]
A big improvement (and would probably take about zero minutes of effort) would simply be  to change "data" to "data (excluding data/media)".  This way, there's disclosure, and in my case, I'd simply have tarred it up and copied it somewhere.  Most people advanced enough to install TWRP probably have the skillset to perform a non-recovery based workaround (or just adb shell into TWRP and copy /data/media to the external sd card.)

Also, WTF...someone else already did the work for this?

**Comment by CaptainThrowback** [comment]
@kitambi If it would take zero minutes of effort, feel free to submit the change to Gerrit for review. TWRP is open-source, after all....

**Comment by Dees-Troy** [comment]
We made this design decision years ago when emulated storage first came about. I'm not entirely opposed to revisiting the decision as emulated storage is widely used now, but there's more to the decision than just, "do we or don't we".

For example, when you restore a backup of the system partition, we wipe the system partition first, then do the restore. If the user decides to restore a backup of the media folder, do we wipe it first? If we wipe it, are we wiping out newer photos, documents, downloads, etc? The whole issue gets a lot more murky when you start to consider these implications.

And no, the average TWRP user can't even spell adb. I mean literally, they can't even spell it. The average TWRP user is running MS Windows. This same user doesn't have adb drivers installed and they don't even know that they need drivers or what a driver is. Our average user doesn't have adb binaries downloaded either.

When the average user makes a backup, they either leave the defaults set or they check every box present in the GUI because more backup is better. This same user won't pay the least bit attention to the implications of restoring a backup of media and how restoring that backup may or may not wipe out their most recent precious pictures of their newborn child.

Let's be real... the average TWRP user doesn't even know how emulated storage works. They have no idea that their /sdcard is really located in /data/media. For a while we tried not creating /sdcard in TWRP for emulated storage devices. Gobs of people complained that their /sdcard was missing and no amount of explaining helped.

**Comment by cmorlok** [comment]
But the same implications already apply when the user restores /data. That
wipes out the latest texts, call logs, whatsapp messages, game credits,...
(unless his apps sync with an online service).

How does the average user deal with that?

**Comment by kitambi** [comment]
I suppose that makes sense when we consider the historical implications.  However, as far as /data/media goes, TWRP doesn't currently wipe /data/media at all unless the user specifically requests it, or they format /data (which I'm assuming is a power user feature, but now I'm losing even more faith in humanity with the knowledge that people are using Android recoveries without having the slightest inkling about what they're doing...also, how do they install TWRP without Android platform tools?)

I think an acceptable compromise, in the immediate, is to add a tag to the partition list that says "excludes /data/media" next to the "Data" name -- after all, if we're talking "average users" here (of which skill level is assumed to be very, very far from a normal distribution), that should get them asking questions.  Also, since "Data" is a pretty broad term, most people ( (set of all users) - (set of twrp developers) - (set of people who lost everything in /data/media) ) expect a backup of "Data" to backup _all_ of their data.

The current behavior is the equivalent of a Windows backup program backing up the entire system except for "C:\Users", or a Linux archive program operating on the entire system, but excluding "/home".  Unintuitive at best, destructive at worst.

I think I'm done adding an "excludes" tag to the partition list, but now need to free up enough space for a test build to run in an emulator. :)  This should allow people who know how to archive their /data/media in a different way, while (hopefully) prompting the "average user" to ask questions about it before they lose everything.

**Comment by JustMyGithub** [comment]
What about aditionally having an advanced option that (de)activates backup of /data/media being offered by the backup dialog. "wipe or move existing data" could be asked whenever a backup which includes /data/media is restored (at the beginning of the restore process).
question still is how to handle Images that include /data/media but have been moved to that very folder.

**Comment by steadfasterX** [comment]
+1 for an advanced "hidden" option for backing up. That means you have to set manually and explicit the inclusion of /data/media. Besides that setting I would see a hint to a FAQ entry online.
That way one can handle new questions if they arrive and update the FAQ accordingly without changing code in TWRP. 

When it comes to restore it have to be the same: an advanced "hidden" option to enable the handling of /data/media again. I would point to a FAQ entry online here, too which explains the impacts goods and bads especially for restoring.

As @Dees-Troy said the average TWRP user will either select all options or simply leave it as it is but my idea is that you have to go into e.g. the "Advanced" Menu and within this you can (de)select the /data/media -> backup and the /data/media -> restore. Before actually starting the backup/restore another hint could be "popped up" like "Danger option choosen - READ THE FAQ!"

To be honest I totally agree that the /data/media backup is a absolutely must have but all users participating here may not are "average users" (creating a github account is a high barrier ;) )  so I believe we should hide it for the average users and give "experts" the option to do so.

**Comment by diedmatrix** [comment]
Just wanted to say that i just lost "all" my data due to this lack of information because i relied on the internet which said "you can FULLY restore your phone!"...... >-< Yes you can doupt me for not being a linux prof. or what else, but this is a huge problem in my opinion! At least there should be an info about this! 

The Folder/Files from "internal storage" visible via Windows Explorer should be under /data/media/0 in the TWRP backup file? This is not the case at my backup - i extracted data.ext4.win000 and data.ext4.win001 with 7-zip and opened and browsed them with 7-zip but no such folder or the folder i am looking for..... :(

Does this mean it wasnt backed up and even if i restore it to the now FULLY WIPED phone, i will not get back my files which where stored under this specific folder?

**Comment by JustMyGithub** [comment]
I am sorry, but I have to tell you that you are right - there is no backup of those data.
Your only chance would be professional recovery - if you did not overwrite the flash memory.

**Comment by leo-bogert** [comment]
@Dees-Troy

> For example, when you restore a backup of the system partition, we wipe the system partition first, then do the restore. If the user decides to restore a backup of the media folder, do we wipe it first? If we wipe it, are we wiping out newer photos, documents, downloads, etc? The whole issue gets a lot more murky when you start to consider these implications.

The solution should be less than 10 lines of code:
Before restoring, show a Continue/Abort dialog which tells the user: "WARNING: This will delete all your existing data including pictures, blabla. Continue?".

> When the average user makes a backup, they either leave the defaults set or they check every box present in the GUI because more backup is better. This same user won't pay the least bit attention to the implications of restoring a backup of media and how restoring that backup may or may not wipe out their most recent precious pictures of their newborn child.

If they don't read the suggested warning message, that's their fault then. Which is a lot better than causing people to lose all their data without it being their fault.

I mean there certainly is no guilt involved in believing "backup" means "copy of all your data". That's absolutely the common-sense interpretation I would expect 99% of people to have.
But there is a certain guilt involved in ignoring a "WARNING" dialog.

So please fix this to ensure data-loss only hits people who are too lax to read "WARNING" dialogs, not the 99% who would assume a backup is a backup :)

Thank you!

**Comment by Natanji** [comment]
I also lost ALL of my data in /data/media because TWRP did not back it up, despite saying with would backup the whole partition.

My phone was broken: recovery still worked, but booting into the OS absolutely didn't. Luckily I could install TWRP and do a backup there, before I shipped off my phone for repair (which replaced the motherboard and therefore the whole storage).

Now everything is just... gone and I'm incredibly sad about it. I had no chance to do a backup differently. The very IDEA of backing up your system via recovery should definitely focus on the use case of a recovery, aka the system no longer booting.

This is a critical oversight and you really have to change this ASAP. You should also tell users whether this will be included or not, regardless of whether this is an advanced option or not.

**Comment by JustMyGithub** [comment]
I agree with you, Natanji and I am sorry for your loss of data.
However your phone could always fail completely out of the sudden, so you should have a backup BEFORE your phone fails.

btw: TWRP 2.8 and above support MTP access to internal storage

**Comment by Natanji** [comment]
It's great if there is MTP access to internal storage, and TWRP still doesn't tell you about this. Now it feels even worse, because just a quick warning message would have prevented this. Right now TWRP completely misleads the user because it SAYS it backed up /data/ successfully, and yet it didn't do so. That is a _critical_ bug.

This bug is two years old and nobody seems to have lifted a finger to even just include a warning message, even though multiple users have reported loss of irrecoverable data. For someone familiar with the codebase, it's certainly easy to just add that. It's outright irresponsible to not just take half an hour to fix this. You created a software that SAYS it backed something up when it didn't, which is one of the worst things a backup feature can do. The software quite literally lies about it, due to a developer choice not made transparent to the user. Great job.

I mean, for me it no longer matters whether this gets fixed. I'm aware of the workings now and would know what to do. I just hope you are aware that all this data loss is the fault of whoever made the decision to exclude a part of a backup without telling users about that in any way. If you can't even invest half an hour of your life to rectify this, you're a bad developer.

**Comment by bigbiff** [comment]
TWRP is open source. Please provide a relevant patch for this issue to Gerrit.omnirom.org. This is not a bug but a feature request so I am closing this issue.

**Comment by Natanji** [comment]
When program says it will do a thing X, and then proceeds to NOT do that thing X, that is objectively a bug. TWRP says it will backup /data, even prints a message afterwards that this process was successful. But it didn't fully backup /data. So this is a bug.

So please, @bigbiff, reopen this issue. I honestly can't see how you can view this as a "feature request". This is nowhere explained in the documentation or user interface of TWRP. So the current behaviour of TWRP is different from what it should do, and from what it SAYS it does. Again: _objectively_ a bug. Any developer knows that.

**Comment by JustMyGithub** [comment]
“Do not exclude“ is a fr. Not Reporting it is a bug, i would agree

Am 10. September 2016 14:56:39 MESZ, schrieb Natanji notifications@github.com:

> When program says it will do a thing X, and then proceeds to NOT do
> that thing X, that is objectively a bug. TWRP says it will backup
> /data, even prints a message afterwards that this process was
> successful. But it didn't fully backup /data. So this is a bug.
> 
> So please, @bigbiff, reopen this issue. I honestly can't see how you
> can view this as a "feature request". This is nowhere explained in the
> documentation or user interface of TWRP. So the current behaviour of
> TWRP is different from what it should do, and from what it SAYS it
> does. Again: _objectively_ a bug. Any developer knows that.
> 
> ## 
> 
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/TeamWin/Team-Win-Recovery-Project/issues/276#issuecomment-246110266

## 

Diese Nachricht wurde von meinem Android-Mobiltelefon mit K-9 Mail gesendet.

**Comment by Natanji** [comment]
"Do not exclude" is one way to rectify the bug, just like reporting about this (right now unreported/undocumented) special behaviour is. It's either make the program do as it says (->do not exclude) or make the program say as it does (->report that the exclusion takes place).
I don't believe opening a new issue about this helps, but if this stays closed then I will do so.

**Comment by bigbiff** [comment]
The reason we exclude emulated storage is for space on the backup storage. Please either use adb pull or mtp to backup storage to your pc. What happens every time you try to backup 15 gb emulated storage to itself? It will fill up. If you have any better way we can review it at Gerrit.omnirom.org

**Comment by steadfasterX** [comment]
The decision to not include emulated storage is ok @bigbiff .
The only thing needed is to TELL the user that this exclusions is there. This is the critical part: Knowledge. If it is known that emulated storage gets not backuped one can prepare.

If someone want the emulated storage to be added as backup source then he/she should do a pull request on gerrit.

For the current implementation in TWRP it simple needs this:

1) on the twrp site (with the hint how to do it manually in general would be perfect. I mean not for every device a howto just a general advice would be enough)
2) in TWRP itself when choosing /data as backup source

I believe I could do 1 and send a PR but not for the second one..
.

**Comment by diedmatrix** [comment]
> The reason we exclude emulated storage is for space on the backup storage. Please either use adb pull or mtp to backup storage to your pc. What happens every time you try to backup 15 gb emulated storage to itself? It will fill up. If you have any better way we can review it at Gerrit.omnirom.org

You should never put your backup also on the internal SD Card! I am doing my backup on the external SD Card and then copy it to the PC. So in this case its absolutely no problem to backup hole /data?!

**Comment by htsmi** [comment]
"Not a bug" is not a reasonable response to the many advanced users who have been misled by the lack of documentation of this unexpected behavior. It is documented in many places that "Internal Storage" is simply a folder stored on the data partition (see TWRP's own FAQ: https://twrp.me/faq/datamedia.html), so I, like many, believed I could reasonably expect that TWRP would back it up when I selected "Data" from the list of partitions to back up.

After all, "Internal Storage" is not listed as a separate partition to back up. So I assumed that by selecting all the partitions that contain user data, I was actually backing up all my personal data. As a result of assuming this, I, also like many here, irrecoverably lost my personal files and photos.

As has been explained by many here already, there are two possible ways to remedy this bug: either improve the interface documentation, or provide an option to include /data/media in the backup (which I personally believe should be the default behavior). Both should be relatively simple for someone who is familiar with the codebase, not an end user like myself. I would consider failing to address this, and continuing to promote TWRP as a backup tool, to be irresponsible at very least.

**Comment by andi34** [comment]
@bigbiff maybe disallow backup to internal storage if /data/media get backuped too?

**Comment by czarsuperstar** [comment]
If people are using a custom recovery they know the consequences. When I wipe I wipe everything but internal and micro SD.
When I back up my stock room is gutted and bloatware free. I back up boot, system, and data and I have everything left as it was before I booted in recovery. That's why every few days I make a new backup. Wipe everything after a backup. Reboot recovery and restore and if everything is fine I delete my previous backup.

**Comment by czarsuperstar** [comment]
Meant to say stock rom and is gutted. ...

**Comment by Natanji** [comment]
@czarsuperstar I don't see the relevance of your comments towards this issue. However, I can tell you for sure that people who use a custom recovery do not generally have the expectation that TWRP leaves internal data unwiped when you tell it to wipe, well, everything. They have to have stumbled over https://twrp.me/faq/datamedia.html for that.

Personally, I have never needed to wipe my phone using TWRP since I only ever used TWRP to root it/install Xposed. The only wipe that happens is when unlocking bootloader, and it always does wipe _everything_ on /data/, it doesn't keep internal storage. So the fact that TWRP does this differently is already non-standard and not what one would expect. It is actually dangerous that TWRP calls a function "factory reset" that does not actually wipe data: imagine that you want to sell your phone over eBay and tell TWRP to wipe the phone. After the "factory reset" you boot it up and see that it seems to have worked because Android gives you the usual setup screens like on first start. You send it off to a stranger but it didn't actually get wiped completely - and all your personal photos and other documents on /data/media end up in the hands of they buyer. Oops! Someone clearly didn't think about the consequences of naming these options properly.

At least on this point though, one can argue that users of TWRP should have properly read the documentation. The backup behaviour that leaves out internal storage though is NOT documented anywhere however. So it's way different anyway.

**Comment by czarsuperstar** [comment]
If someone sells a phone with a custom recovery than I would not trust that person. My comment is geared towards advance users and development. We are discussing the topic of a custom recovery which SHOULD NOT BE USED BY BEGINNERS. I read and continue to read. Instead of waiting on the T.W.R.P. Team to decide if they want to Dev for my device I compiled my own. It takes ten minutes to literally. Sorry if I was misunderstood.

**Comment by Natanji** [comment]
You can restore the stock recovery afterwards so eh.

Also it's nice that you can compile your own things in ten minutes. I can't and TWRP is used mainly by users who cannot do so either, I'm sure. Your elitist comments contribute nothing of value to this topic at hand. We are discussing a bug here so please get back to topic or get out.

**Comment by czarsuperstar** [comment]
I have never ever seen a post telling users to wipe internal storage so when I wipe cache and dalvik my internal is still in tack. I have been on the scene since the G1 so whoever deletes their internal storage that's their fault because it's stated the damage done to your own device is your responsibility point blank and also that I don't have that problem.

**Comment by jcadduono** [comment]
If a patch is submitted for this, it would probably be best to simply have "Internal Storage" (/sdcard, excluding /sdcard/TWRP) as a backup option, it should not show up without the RECOVERY_SDCARD_ON_DATA flag - at least if users see Internal Storage in the backup options, they'll perhaps understand that "Data" does not include it. For devices with internal storage outside data partition, they can of course create their own backup=1;display="Internal Storage" line for /sdcard in fstab.
This feature would really only be useful to users with OTG storage or microSD's though, I guess.

**Comment by prataczak** [comment]
I understand the reasoning behind not backing up the folder with all of your backups. So why not provide an advanced option to backup data/media exclusively to an external SD card? This way, your oblivious 'average user' can go on thinking everything is backed up, even though it isn't, and the people who know better are happy.

Obviously this isn't just 'a few people.'

There is a large group of people who turned to TWRP because they are using custom ROMs. This is the group of people who are most at risk of losing their data, because they will possibly/very likely revert to stock at some point, whether it's permanent, or simply to update the profile/PRL (Sprint users have no way to do this with a custom ROM).

Saving 'new photos' from being wiped is a non issue with apps like Google photos that automatically back up all of your photos and videos as soon as you take them.

**Comment by Natanji** [comment]
Advanced users that customize their phone often don't _want_ to back up all
their data to the cloud. TWRP backup is not really needed when you use
Google's backup solutions already. But I for one don't want Google to have
access to my private photos. So don't paint this as a non-issue.

prataczak notifications@github.com schrieb am Mi., 21. Sep. 2016, 09:29:

**Comment by prataczak** [comment]
You know, come to think of it, I think I started using Google photos to back up my photos _just because_ I got screwed by TWRP a year or two ago and lost some photos when I went to revert to stock to update my profile and PRL.

Not trying to start a debate, but you really should get over the notion of "google having access to your private info." I'm not saying that privacy doesn't matter. I'm simply pointing out that the government, and the NSA have been keeping tabs on you way before you were aware of it. There's no respect there, and there certainly isn't going to be any legal constraints stopping them from having power and authority that no one gave them.

But also... unless you're taking naked pics of your girlfriend... who cares? Honestly who cares if google can see pics of your cats and your kids' birthday parties. And if you are taking some pics you want private, you can turn off the automatic uploading feature.

It used to make me uneasy too. But then I realized in the grand scheme of things, Google having access to your 'private photos' (as if they even have the resources or the inclination, I'm pretty sure they have better things to do) isn't what you should be worried about.

**Comment by czarsuperstar** [comment]
I am appreciative of the efforts of the Team.  As seeing we are limited in the choice of recovery.  I didn't have any knowledge on building a recovery but I read and read and read some more.  Tell CWM To bring their recovery oh I forgot the recovery is not updated.  I learned to build a recovery in a month from Dees from the XDA forum. Let's try to find a solution instead of demand. How many phones out there started from this team? Not all. Someone took up leadership and got in the trenches. Let's find a solution instead of bashing because the time being taken on here is time you could solve your own problem.  I asked for help from a lot of people and have received none. I took various configurations from other phones and my recovery is solid. Thanks.

**Comment by steadfasterX** [comment]
I added an explanation of the backup behavior to the twrp website which is the first step: **information**.
This has to be reviewed first but hopefully gets added public soon.

The second step should be to build a solution for this so both can be happy: the one who wants to backup /data/media and the others.

@czarsuperstar: thanks. really. 
Many people do not understand how open source works!

**Comment by czarsuperstar** [comment]
@steadfasterX No problem. But it's getting annoying for the so called advance rooter Lol to demand solutions with logs and all but can't fix the so called problem. In the beginning there was no internal storage. That's where apps 2 SD started being able to put apps and frameworks on a SD ext. Basically stating that open sources are for learning and contributions.

**Comment by steadfasterX** [comment]
**Online** :)
Sorry for any grammatical and some format issues there (will be fixed in a while)

Please see:
https://twrp.me/faq/backupexclusions.html

**Comment by leo-bogert** [comment]
@bigbiff 

> TWRP is open source. Please provide a relevant patch for this issue to Gerrit.omnirom.org. 

Actually since Sept 4 2015 there have been 4 commits and a pull request linked against this issue, scroll up. They're on a fork of TWRP so they should be easy to port.

> This is not a bug but a feature request so I am closing this issue.

How is dozens of people coming here telling you they lost their data not a bug?
And even if it is not, then label it as "feature". I don't care what it is called. Just please fix it whatever you call it :)

> The reason we exclude emulated storage is for space on the backup storage

Then check whether "getFreespace() > expectedBackupSize() \* 1.10", and abort with an error if not.
Should be less than 10 lines of code which can save users losing years worth of memories (= their camera pictures).

> Please either use adb pull or mtp to backup storage to your pc. What happens every time you try to backup 15 gb emulated storage to itself? It will fill up. If you have any better way we can review it at Gerrit.omnirom.org

adb is not an option because it leaves out empty directories and kills the file timestamps.
MTP is so new that it doesn't work properly on many Linux machines, it doesn't work on mine.

Besides let's not lose ourselves in details here:
The big issue is that the "backup" feature is not a backup because it leaves out arbitrary data.
This has caused probably a dozen of people to lose all their data, and 57 comments in this bugtracker issue. Given the large barrier of entry for users to actually consider registering on GitHub, I would say this is more than enough proof that lots of your users are unhappy about this.
If it makes so many people unhappy, why not just change it?
Sure, you can tell them to jump through hoops instead, but that won't make the fact go away that it does seem to annoy them a lot :|

Sorry for joining in with this brigade, but from a neutral point of view without considering the technical details this is really obvious: If something says it backs up a directory called "/data", it shouldn't leave out _any_ files or directories inside this directory. This doesn't depend on whether the thing is a phone, an "emulated storage", an sdcard or whatever. It could have been about some 1970s Unix mainframe as well. If you say you back things up, then please back things up.

**Comment by czarsuperstar** [comment]
I have never lost data stored on internal storage unless I wiped. People are demanding a fix. Demand other recovery makers to update their sources. I really don't like the demands being made. Compile your recovery and add this feature. Check my BoardConfi.mk since this feature is my build because I back up data and internal is not enough. If you were knowledgeable to root your device make a recovery. Im finished.

**Comment by JustMyGithub** [comment]
Everybody reading here knows about this issue. Telling people that already KNOW about this issue to compile their own recovery with this warning included simply misses the point.......

**Comment by czarsuperstar** [comment]
All I'm saying is research. I have never built a recovery.  I downloaded Omni from the Xda forum for compressed sources google : compressed cm13 sources Xda. Read necessary information and I built. There is roughly 9 members of Team Win. There's over a surplus devices out there. Most of the recoveries built are from other people who decided to not take the easy route and wait. There's instructions on Xda to compile sources. Let's all work on this problem together.  I have been trying to fix this issue.

**Comment by SarenT** [comment]
I have lost my data as well. I am really angry because of the lack of the warning (I am also thankful for the software though!). It is just sad, these things need to be reliable and it is counter-intuitive to not back-up /data/media. Some people do not want to backup their private data in Google and this is usually out of principle and not actual fear. Maybe noone cares about my information but that doesn't change anything. At least a warning would have been enough.

**Comment by smemsh** [comment]
61 comments, 26 participants.  So many people affected by this bug, new people coming in all the time.  Closing the ticket just because of some whining users, and pretending it's not a real bug is irresponsible.  Maybe nobody ever fixes the bug, nobody forcing you to write code but failing to acknowledge data loss from continual stream of users affected by the issue is just being stubborn.

**Comment by prataczak** [comment]
They just don't want to take the time and effort into coming up with a solution that isn't going to affect the majority one way or the other.

That's what this boils down to.

**Comment by bigbiff** [comment]
It's not a bug. TWRP for years has been designed to skip storage from backup. It's just a design decision. If you want to backup storage use adb or mtp. Or plain cp to external_sd.

**Comment by blackwind** [comment]
Even one user losing all their data is one user too many. Justify it any way you'd like, but choosing not to react to this ongoing parade of digital tragedy is objectively irresponsible through and through.

**Comment by zeorin** [comment]
@bigbiff Can the menu make it clear what is excluded when backing up /data?

**Comment by steadfasterX** [comment]
Adding a hint will make absolutely no sense if ppl can't read which is the case just here: 

As stated some posts before I provided a clear description about all this here: 

https://twrp.me/faq/backupexclusions.html 

Yes another hint when backing up pointing to this FAQ would be great but if ppl do not read when pointing them.. 

You have 4 options at least: 

1) read the faq and respect the decision 
2) read the faq and go away and use another recovery 
3) read the faq and offer a solution in terms of code for adding a hint when / before backing up / restoring
4) read the faq and offer a solution in terms of code for adding backing up /data/media as an option the user can choose

Got it? The first thing to do is read the faq.

**Comment by SarenT** [comment]
I am sorry but if someone is using an app/piece of software, a crucial information such as this, should be available on the user interface. Many simple users are willing to drop manufacturer shitware and Apple-turd products to move to open source solutions. Cyanogen Mod, Ubuntu, LibreOffice are wonderful examples. And it is a GOOD thing. But when maintainers are simply unable to answer the question "why can't we simply put a small warning inside the GUI?", then the discussion is over. I simply don't accept steadfasterX's answer.
Many more people will read it if you put it on the GUI! Most of the people are too lazy to read the FAQ because of the fact that a normal user would expect any crucial information/hint would appear on the GUI. Similar discussions are on OnePlus forums, StackExchange etc. FAQ and documentation is for people who wants to learn more about it.

**Comment by andi34** [comment]
If the user are lazy, they should not flash anything on a phone or tablet. You can easily brick a device while not reading.

**Comment by SarenT** [comment]
No that is not true. Flashing is risky? yes. Is it an easy process to perform? with the current tools available, yes. So even a lazy person should be able to do it. When it comes to computers, anything should be possible by following instructions. A GUI provides those instructions. I am sorry but not everyone has the time to dig into the deepest pages and spend hours in docs and FAQs. Because if it is not this issue, then it might be something else. So the only proper way of using a software becomes according to your point of view, reading the whole documentation (just in case) before using anything.
I hate laziness, when there is no other solution. But again all users here demand is a simple warning, because it leads to major and critical misunderstanding.

**Comment by htsmi** [comment]
Unsubscribing. Numerous users have sought out this thread to let devs know they have accidentally lost data this way; devs have made it abundantly clear they do not care. Not wasting any more of my time on this.

**Comment by andi34** [comment]
@htsmi as you noticed: it is an opensource project - if something bugs you, feel free to add it and push it for review. Thats the way OpenSource works.

**Comment by steadfasterX** [comment]
I told it **SEVERAL** times: If you don't like the way it is atm just **change** it and send a pull request to the twrp gerrit. 

**Open Source works exactly this way (I have to say that over and over again).**

If you feel uncomfortable with the current situation **change it and stop demanding**.
If you don't want to change things and you still uncomfortable with the current situation - go to a developer and give him enough money to build your own recovery including support and adding features as you like.

and btw: 
No one (90% of ppl) reads any hints. No one will actually read a warning even when within the GUI. This will not solve anything. **jM5C**!

**Comment by leo-bogert** [comment]
I've volunteered in open source projects for years, and the people who say "it's free software, you cannot demand anything from us because we're volunteers" are missing an aspect IMHO:

Imagine you repair the car of a friend in need who cannot afford paying a mechanic.
Then suddenly they have a dangerous accident because the brakes didn't work properly.
Would you tell them "Yeah I did know the brakes are broken, but I was doing it for free, you cannot expect me to take care of everything, I ignored the broken brakes!"
No, for sure not. You would have fixed the brakes or told them to have someone else do it. Doing it for free doesn't imply that you would like to impose danger upon other people's stuff.

Now just because we're not personal friends of yours but strangers from the Internet, does this mean it suddenly is OK to release a dangerous piece of software?

I believe volunteers should as well follow basic work ethic of not releasing something into the wild which can cause damage to other people's stuff.
If you release free code that is a big service to humanity and I am thankful for that; but if you don't have the time to fix dangerous bugs then please rather don't release anything at all.
You're of course not obliged to write any more code for TWRP at all, but if you're still maintaining this project, it would be very polite to fix this as the next thing you do - whenever it may be, even if it never happens. Just please don't do other stuff first.

Please trust me that I know that volunteering is hard and honorable work; I've done my fair share of it.
And whenever I do it, I also try to follow the same quality standard as if I was doing it for my friends or being paid or whatever :)

**Comment by steadfasterX** [comment]
Thats not the way open source works.

**Comment by yucandu** [comment]
Yeah this is definitely confusing.  For years people on the XDA forums said "Do a nandroid backup with TWRP, it backs up _everything_", and TWRP certainly appears to make it looks like it is backing up everything in its UI, but it's not.  There are things it goes out of its way to warn you about, like having SuperSU installed or whether a file system is mounted - things I don't really care about and can figure out myself - but warning you that a TWRP backup isn't a complete backup is apparently a really controversial issue that would take hours of someone's time and isn't worth it and has people in this thread fighting with each other and getting raised blood pressure over?

**Comment by jxu** [comment]
I lost all my Screenshots, Downloads, probably lots of important stuff in /sdcard that I don't even know about... Geez I might even have to dig through old PhilZ Touch or Clockworkmod backups

The mentality is strange: "Go fix it yourself" while TWRP proceeds to work on something else presumably far more important than this issue that many users have experienced? The worst part is reading about people who've lost far more than I have. Tragic, really

**Comment by CollinChaffin** [comment]
I share everyone's pain here, I too just got burned and lost valuable data thinking these "full partition" backups I've been friggin waste hours every week to take my Android devices offline boot to recovery to backup would be worth a #^@& and 100% of them are 100% worthless to me (on almost a DOZEN different Android devices in my home) because I thought TWRP was imaging every partition on the device.

You know as much as I am usually an Android evangalist, this is where Apple is KICKING ANDROID'S ASS.  You see stop for 2 seconds to think of this just as a human being holding a device in your hand.  Apple has put one thing as the #1 most important for BACKING UP this device in my hand......MY FRIGGIN DATA.  Screw the OS, screw the device, ,MY DATA.   Pictures of my kids.  Voice mail.  SMS.  When I go to restore any of my Apple devices, you know what?  My friggin data is there when it completes - every time.  Every picture, voicemail, SMS, email.  How many did TWRP save for me?  ZERO.

I just wonder what total genius developer took it upon himself to decide that the user data is the ONE FRIGGIN THING we WONT BACKUP, we don't touch, and frankly WE AS ANDROID DON'T CARE ABOUT.  When we go to restore this device in your hand.......you can drop dead cause as Android we care less about you, the owner.  As long as our precious DEVICE is okay, screw the rest of it, including any of your data.

If you read this thread, the resounding theme is......these folks seem to, like me, disagree with the device meaning a damned thing.  As far as I'm concerned, if my kids' pictures, my voicemail/email, SMS isn't there when I ask TWRP to save my life and restore my data.......screw the device it goes into the TRASH CAN.

This issue IS the poster of one huge drawback of open sourcing this stuff......you take away the management control to FIRE young developers who even think of coding a backup utility that doesn't put the end-user pushing the button as the #1 person to satisfy with the software operations and putting a seasoned developer in charge to actually take the time to roadmap such a critical utility and gather input from those who actually have to digest the end product before just deciding on our behalf as to what is important to us.

In Teamwin's defense, it's Google that should have been stopped from ever releasing an OS of any device without any reliable user data backup and restore function - and do NOT tell me the piddly Google provided backing up my username/password and <maybe> a few apps data IF they code themselves to use Google - that is NOT a valid device backup/restore.  Every Android install should have a full user data backup/restore function just like Apple and it should have happened almost a DECADE ago now......try explaining to your family members why Android is so much better and the almost invisible protection they had on Apple - that NOT having that on Android is somehow okay.  It's not.

Does anyone know (I am just starting to dig through all the forks) if some DEV team has already forked to add the ability to backup and restore 100% user data, like Apple and every other backup/restore system on the planet does?  I'm guessing since I didn't see it mentinoed here, that the answer is no, but thought I would ask.

**Comment by JustMyGithub** [comment]
This was not a single-noob decision obviously, as in that case this "bug" would have been "fixed" long ago.
The philosophy of adb backup is to grab what the user cant. The user can backup the user data manually (mtp)
TeamWin seems to share that philosophy, no matter how many people loose gigs of pictures because of this.
Even though this is the case, there should be a red warning message....

**Comment by Dees-Troy** [comment]
Maybe this will make some people happy:
https://gerrit.omnirom.org/#/c/21635/

**Comment by steadfasterX** [comment]
Lol ;)

**Comment by jxu** [comment]
If it helps anyone, most recently deleted images and videos can be recovered with file recovery apps. It's best to do this as soon as possible.

**Comment by JustMyGithub** [comment]
This is only part of the story, wxv!
You are right, but usually you notice this after you restored a backup and maybe long after you deleted the data (as you think you have a backup). On most devices the internal storage is just a folder on the data partition, not a separate partition. In this case during the restore you already wrote a hole bunch of data to the very partition which you want to recover data from.

Of course it is worth a try anyway.

**Comment by steadfasterX** [comment]
@CollinChaffin 
no need to dig through the forks ;) 
See this (untested) pull request by @eugenesan (you can find it linked to this issue as well... some pages above.): 

https://github.com/omnirom/android_bootable_recovery/pull/39

The way is not that perfect because you have the choice on build only but nevertheless it should do it. 
Unfortunately the creator hasn't send this pull request to the official gerrit therefore it never get reviewed.

**Comment by jcadduono** [comment]
yeah that's a pretty awful solution - the right way to do it would be to separately include Internal Storage in the backup page, and exclude its own backup folder, as well as display a warning page if you are backing up to internal storage. (to alert the user that if they intend on wiping internal storage, they will lose the backup they are creating and to instead select another storage or swipe to proceed)
by having Internal Storage included along with Data in the options to backup, it should easily imply that Internal Storage is not included in a Data backup.
it's the only way it will get accepted.

**Comment by zeorin** [comment]
@Dees-Troy https://gerrit.omnirom.org/#/c/21635/:

> Maybe this will make the people in issue 276 happy, but probably not because they already lost their data, or because the warning text is yellow instead of red.

It makes me happy, at least. You even got a chuckle even though I lost all my data.

**Comment by dunqan** [comment]
It is devastating. It took 3 years of dev trash talk and arguing if it is a bug or fr... instead just accepting to change one string resource in code... I'm sorry for any offense, but I'm not the noob user, I did know that internal storage is part of data partition and I was sure I had my data backed up and safe, because I've backed up the data partition to external sd card, copied it to my PC and backed up once again to cloud backup.
Unfortunately, as many before me (and probably many still not aware of it) I was wrong. Simply because "Backup data partition" meant "Backup only portion of your data partition".

**Comment by jxu** [comment]
Well, this is what happens when you write third-party code for developers (or more likely enthusiasts) and not for end users - I can at least say my blind faith in the organizational ability of open source has been humbled.

**Comment by chrcoluk** [comment]
wow the attitude of twrp devs here is unreal.  They coming across as very lazy refusing to add a couple of lines of code, I agree 100% with the promise that TWRP is misleading, it either needs to include data/media or display a warning that /data is not a full backup of data.  Doing the latter should be extremely easy for a twrp developer.  But this type of response is not surprising when you have a monopoly which TWRP does following the death of clockworkmod.

**Comment by CaptainThrowback** [comment]
@chrcoluk Not sure why you're still commenting on this now, especially when this change was merged into the latest version of TWRP?

**Comment by that1** [comment]
Wow the attitude of some users here is unreal. While I agree that this was a big usability issue and it should not have been closed last year without any code change, that code change is now in 3.1, so there should be no more reason to complain. Also TWRP is open source and anyone can improve it, but for some reason people whine about "monopoly" instead of submitting a patch.

**Comment by jant90** [comment]
+1 please fix, I just found out none of my photos and other data weren't backupped and are now lost forever after making a backup of all partitions to an USB-OTG device.

**Comment by transplier** [comment]
I know this issue is closed, but I just wanted offer my two cents to devs (for any project) reading this - usability is important, even/especially in technical tools. Too often (myself included) we assume the user knows what we know and will use the tool in the same way as we intend. We may even see misguided users in a negative light instead of feeling empathy. I'm guilty of this.

It's now clear to me that TWRP is for messing with the android system, not user data (kindof). This wasn't clear to me a day ago. One of the words in the project title is _Recovery_. I'm a technical user. I've built custom Android ROMs from source and messed with firmware. I know how sdcard emulation works and why it was implemented. _I still lost data._ I didn't read the TWRP source or verify my backups. I didn't even read the FAQ on twrp.me. Shame on me.

However... a _lot_ of users (myself included) would have been saved from catastrophic data loss had a little extra thought gone into the feature. Instead of being satisfied with the tool and potentially contributing, we now have users (even in this thread) which now see the ecosystem in a negative light and are actively pushing people away. This hurts all of us. By all means, adhere to your design philosophy. Just recognize that your philosophy is not universal. Be explicit, not implicit.

Maybe I'll try my hand at implementing a "full" backup in TWRP. Would sure be nice to have a one-stop shop to backup everything with one button. I really wasn't planning on writing rsync scripts today, but so it goes.

**Comment by jxu** [comment]
@progman32 It's funny that you mention rsync, since I wrote a quick-and-dirty rsync script just because I wanted a semi-automated way to copy over my data folder.

**Comment by transplier** [comment]
Yeah, as far as backup tools go rsync seems to be the one I always settle on too. Mind sharing the script for folks in this thread? Do you run it from within the system rom or within TWRP?

Edit: I just used the binary [here](https://github.com/floriandejonckheere/rsync-android), formatted an external disk as EXT4, booted TWRP, mounted it at /usb-otg, and ran a simple `rsync-arm -av /sdcard /usb-otg-2`. That worked, so then I implemented incremental backups like [this](http://www.mikerubel.org/computers/rsync_snapshots/).

Works well enough for me. Sorry for the OT.

**Comment by eltznth** [comment]
I'm very surprised that the devs still insist on making the decision for us whether or not we get to back up or restore /data/storage. I wish they didn't go out of their way to make a "feature" that excludes it in the first place. Now that they've gone out of their way, maybe they should list it separately with their desired defaults so the user can choose to back up or restore /data/storage or not. 

Anyone know how this open source thing works well enough to make a fork to fix this "feature"?

As a work-around, would it work if you renamed /data/storage to /data/somethingelse before backup?

**Comment by EliteMasterEric** [comment]
I finally arrived here after a long search through my old phone backups looking for my old game saves.

I was under the impression that what TWRP provided was a full backup of my phone and its files, and instead it only saves random system files. On top of that, the development team claims to have "resolved" it by adding a little warning message, but too little too late my game saves are gone.

What's preventing you from just adding a checkbox that backs up the /data/media folder on the phone? If people want to back that up through other methods they can.

EDIT: Apparently literally nothing is preventing you from adding a checkbox because pull request omnirom/android_bootable_recovery#39 exists

**Comment by T3rm1** [comment]
I have just lost tons of files because there was no hint that TWRP does not backup /data/media. I made a backup and wanted to restore that after a failed encryption of my phone. Of course all files in /data/media were encrypted and I thought after restoring my backup everything is fine again. Wrong. All data is lost because the data is still encrypted.

Can someone explain to me please why the folder is not backed up?

**Comment by Natanji** [comment]
Read this thread and the linked explanation in the TWRP documentation.
Also, all current versions of TWRP warn you that /data/media is not backed
up - did you use an outdated version?

T3rm1 <notifications@github.com> schrieb am Di., 27. Juni 2017, 11:56:

**Comment by SarenT** [comment]
Probably people did their backups years/months ago. Or they are not aware of the issue, until they need to restore at some point. While their TWRP version remains outdated.

**Comment by steadfasterX** [comment]
at least for the issue of outdated TWRP versions there is the official TWRP app avail which makes that easy including notifications (as long as the TWRP version is official and the user installs and use the app ofc..)..

**Comment by T3rm1** [comment]
I read the link and it makes no sense at all to me why the folder is skipped. What is the best way to back up the folder at the moment? Are there any "improved" versions of twrp?

**Comment by steadfasterX** [comment]
e.g. with the [flashfire app](https://play.google.com/store/apps/details?id=eu.chainfire.flash).

**Comment by zhou13** [comment]
Why I still cannot backup internal storage?  What's the problem of adding a new checkbox "Data partition WITH INTERNAL STORAGE"?

**Comment by rernko** [comment]
Just spent all day trying to figure out how to make a proper backuo of /data/media. Hopefully this makes sense:

`adb forward tcp:7080 tcp:8080 &&\`
`adb shell 'tar -zcvf - /data/media | nc -p 8080 -l 1>/dev/null' &\`
`sleep 1;\`
`nc localhost 7080 > media.tar.gz &&\`
`adb forward --remove tcp:7080`

**Comment by jcadduono** [comment]
@rernko interesting way to accomplish this, have you tested it with >4GB stream?
I used to just do `adb exec-out 'tar cf - /data/media | gzip' > sdcard.tar.gz` but always ended up with ADB timeouts or failures near the 4 GB mark, I assume netcat gets past that issue, nice way around it

**Comment by rernko** [comment]
The resulting 11.1 GB archive seems ok. Haven't tried putting it back tough...

**Comment by nailyk-fr** [comment]
This 4gb problem looks like a computer fs problem. Can you give a try with some ext4 or f2fs fs? 

Le 20 juillet 2017 21:36:09 GMT+02:00, James Christopher Adduono <notifications@github.com> a écrit :
>@rernko interesting way to accomplish this, have you tested it with
>>4GB stream?
>I used to just do `adb exec-out 'tar cf - /data/media | gzip' >
>sdcard.tar.gz` but always ended up with ADB timeouts or failures near
>the 4 GB mark, I assume netcat gets past that issue, nice way around it
>
>-- 
>You are receiving this because you are subscribed to this thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/TeamWin/Team-Win-Recovery-Project/issues/276#issuecomment-316807624

-- 
Envoyé de mon appareil LineageOS avec K-9 Mail. Veuillez excuser ma brièveté.

**Comment by jcadduono** [comment]
@nailyk-fr errrrrrrr, do you figure i'm running windows 98 with fat32? lol
testing now it actually looks to be completely random where adb exec-out times out, anywhere from 60MB to 2GB, and when it does I can't reconnect...this is a completely different issue from the thread though and is probably just an adb server/client daemon issue :/

**Comment by transplier** [comment]
You can also consider using rsync and an external USB drive. See my post above from April 20.

**Comment by nailyk-fr** [comment]
It wont compress but
adb pull /dev/block/platform/*/by-name/userdata .
should work. 

Nice netcat trick !

**Comment by eltznth** [comment]
I came up with a good solution so I hope the devs don't screw this up in the next release. The solution is to rename /data/media to /data/media123 or whatever prior to backup. After backup or restoration, remember to rename /data/media123 back to /data/media. 

Easy or what?

**Comment by eocanha** [comment]
The script to restore the media.tar.gz backed up by rernko's netcat script (thanks!) would be this one:
```
adb forward tcp:7080 tcp:8080 &&\
adb shell 'nc -p 8080 -l | tar -zxvf -' &\
sleep 1;\
nc localhost 7080 < media.tar.gz &&\
adb forward --remove tcp:7080
```

**Comment by WLF0X** [comment]
> I came up with a good solution so I hope the devs don't screw this up in the next release. The solution is to rename /data/media to /data/media123 or whatever prior to backup. After backup or restoration, remember to rename /data/media123 back to /data/media.

How do I know if it still works? Only September build?

**Comment by rernko** [comment]
Put it to the test! (And post results here 😉.)

**Comment by mikhoul** [comment]
I lost everything in my "/data/media/" and I'm far than being a newbie with computers and programming, 99% of the review of TWRP said it backup **EVERYTHING/Mirror Copy** to don't worry only to find that it's not true  and you lost many precious files 😞 .

Most casual/regular users don't know what is "/data/media/" when they use TWRP they only want like any general full backup solution to have it to backup **ALL** the files and **NOT** all the files except some.... it make no sense.

**Even the warning mean nothing for 70% of the user since they don't really know what is "/data/media/".**

The current behavior is the equivalent of a Windows backup program backing up the entire system except for "C:\Users", or a Linux archive program operating on the entire system, but excluding "/home". Unintuitive at best, destructive at worst.


**Why being so STUBBORN and don't add the OPTION to backup EVERYTHING including "/data/media/" ?** 🤕

**Comment by bes1002t** [comment]
The main purpose of an Backup is to save all data especially Photos and Videos! That TWRP does not include /data/media means that the backup mechanism of TWRP is not capable to create Backups which fulfill the main purpose of an Backup.

Please implement a way to include /data/media.

**Comment by CaptainThrowback** [comment]
It's open source. Anyone can submit a patch for review for this. Please add me as a reviewer when you do. Thanks.

**Comment by steadfasterX** [comment]
It's in the works.. I was sick of not having it and no - I am **not** a TWRP developer and even **not** a CPP one and still.. its opensource, I love opensource and so I contribute like everyone could do..

https://github.com/steadfasterX/bootable_recovery_twrp_fork/wiki/Internal-Storage-(data-media)-Backup

as said its in the works so may need a lot of work before merged but I really hope I can achieve this one day.

**Comment by leo-bogert** [comment]
@steadfasterX:
Thanks for being willing to deal with this!
Are you aware that above in this issue, on Sept 9, 2015, someone referenced it from commits which likely implement the necessary code You could have a look at those commits to spare you some or all work.

**Comment by steadfasterX** [comment]
> @steadfasterX:
> Thanks for being willing to deal with this!
> Are you aware that above in this issue, on Sept 9, 2015, someone referenced it from commits which likely implement the necessary code You could have a look at those commits to spare you some or all work.

No I have re-implemented it from scratch..
and tbh the main parts are done and working fine already (as you can see at the linked wiki page) and only some smaller bugs are open but keep in mind:
*  there are still a lot of tests to do (which all can cause delays as things need to be implemented)
*  even the actual code merge can take time as it must fit the needs of the TWRP lead devs ofc.

**Comment by beerisgood** [comment]
Maybe work together with https://github.com/kdrag0n/tipatch ? 
@kdrag0n

**Comment by steadfasterX** [comment]
Basically his tipatch was the reason why I started it right now. I don't wanna say it's bad how tipatch works but.. Let's say it's a complete different approach and works completely different... and is imho far away from being ideal.. 
Don't get me wrong I love to see ppl find solutions which work! I do the same often enough but it's not the way I wanna see it in TWRP..

**Comment by kdrag0n** [comment]
Author of Tipatch here.

I actually ended up making Tipatch after losing data and commenting back in 2015. After watching this issue, I was disappointed that the upstream TWRP developers did not want to add an option or something to include internal storage Instead, they just added a warning.

I agree that my solution is not ideal. It's meant as a stopgap until upstream changes their mind or a major fork arises with the feature. I've considered forking and adding this feature as a seperate "Internal Storage" option on the Backup screen, avoiding the issues of Tipatch's binary patching. Let me know if anyone wants me to try that.

@steadfasterX I don't think adding a setting for it is a very good approach because it may not be obvious to first-time users. Since "Internal Storage" is already used for /data/media everywhere in TWRP, we should show a backup option with that name.

Edit: sorry about the spacing, caused by using email.

**Comment by CaptainThrowback** [comment]
@kdrag0n did you look at the Wiki link and the screenshots? That's what his solution does...

**Comment by kdrag0n** [comment]
@CaptainThrowback Just looked at the wiki link, it was showing up blank in my client. The point still remains that it shouldn't be under a Settings option because it's already an optional checkbox in the backup partition list.

@steadfasterX another comment: `/data/media` shouldn't be forced as vfat because it's not all the same owner and group. `/data/media/obb` contains media_obb:media_obb, and `/data/media/0/Android/Data` contains media_rw:30013 (example, group is determined by app).

**Comment by steadfasterX** [comment]
@kdrag0n 
Thanks for sharing your thoughts. 

I have removed the hardcoding to vfat already (I just forgot to update that wiki part) even when it really doesn't matter as my solution will never format anything. Anyways that's reverted and gets auto detected as before. 

The decision to enable that setting or not as a default can be taken by the TWRP maintainer with a build flag. I believe that this is the most flexible way to handle it. Yes ofc the behavior what happens when the build flag is not set is to NOT enable that setting and this can be discussed. 

For the moment it will stay like this and maybe one day when all have been thoroughly tested on several devices we just switch the default to enable it.

**Comment by kdrag0n** [comment]
@steadfasterX I see, thanks for explaining. Good luck on getting it merged into upstream.

**Comment by androidacy-user** [comment]
I'm thinking maybe to address the issues presented with Internal storage a couple points need addressed:

- Firstly, unlike system, we should consider a recursive merging strategy for restore instead of erasing first. Unlike system, this is significantly less likely to cause catastrophic issues when restored. In addition, we would never overwrite newer files (needs consideration).
- A recursive merging strategy should be also considered in all /data backups. Perhaps a UI setting to "Merge userdata instead of format". 
- Currently some devices use persist for settings instead of data. Provided the persist partition is big enough, maybe it should be default
- Don't exclude /data/TWRP but /data/TWRP/backups. Rationale: it's actually nice to have a backup of TWRP settings and/or theme
- Bare minimum data backups should include /sdcard/Android - some apps store important data there
- Backups of data should have a check mark box under data for internal storage perhaps [✓ Internal storage - includes Downloads, Pictures, etc]

Quite honestly I've yet to find a convenient way to backup everything internal storage - MTP and most apps don't see the contents of /data/media/0/Android/*/<other apps>. And heaven knows a NANDROID backup is probably the safest and most complete backup method - complete except for the one unpleasant surprise a lot of users will get when restoring backup.

As for @Dees-Troy suggestion that most users don't even know what /data/media is, you are EXACTLY right - which is _precisely_ why this should be included by default in data backups. Most people don't read data as "/data excluding /sdcard" but "my data". Although UI has been improved to reflect this, there will still be some misunderstanding. 

Personally I have everything I care about backed up to my servers but many don't. Many if I wiped their phone today several years of history would be wiped alongside it.

Just thought I'd put my two cents in an important issue that's been dead for over a year now. Hope it means something to someone

**Comment by Manwithpants812** [comment]
Complete clownery that users aren't AT LEAST given the option to do this, still 10 years later, and the misleading nature of its omission leading to people losing their internal storage/media. If you're so apprehensive about "what about the new /data/media already there", simply add an additional confirm screen that spells that it will delete your existing internal storage in case of restore. Having it be backed up as an option is invaluable. Users use TWRP specifically to backup their device, that means the whole thing and they assume so.

Backing up to OTG is my only option and I don't have an intuitive easy way do this for my internal storage via TWRP.

**Comment by CaptainThrowback** [comment]
> Complete clownery that users aren't AT LEAST given the option to do this, still 10 years later, and the misleading nature of its omission leading to people losing their internal storage/media. If you're so apprehensive about "what about the new /data/media already there", simply add an additional confirm screen that spells that it will delete your existing internal storage in case of restore. Having it be backed up as an option is invaluable. Users use TWRP specifically to backup their device, that means the whole thing and they assume so.
> 
> Backing up to OTG is my only option and I don't have an intuitive easy way do this for my internal storage via TWRP.

Awaiting your patch on Gerrit to enable this, thanks for your concern!

**Comment by jxu** [comment]
Wow what a throwback from that notification. I'm pretty sure this project is long dead.
